src/AppInstallerCLICore/Commands/UpgradeCommand.cpp

```diff
@@ -17,16 +17,66 @@ namespace AppInstaller::CLI
 {
     namespace
     {
-        bool ShouldListUpgrade(Context& context)
+        // Determines whether there are any arguments only used in search queries,
+        // as opposed to listing available upgrades
+        bool HasSearchQueryArguments(Execution::Args& execArgs)
         {
-            for (Execution::Args::Type type : context.Args.GetTypes())
-            {
-                if (type != Execution::Args::Type::Source && type != Execution::Args::Type::IncludeUnknown)
-                {
-                    return false;
-                }
-            }
-            return true;
+            // Note that this does not include Manifest (no search) or source related args (used for listing)
+            return execArgs.Contains(Args::Type::Query) ||
+                execArgs.Contains(Args::Type::Id) ||
+                execArgs.Contains(Args::Type::Name) ||
+                execArgs.Contains(Args::Type::Moniker) ||
+                execArgs.Contains(Args::Type::Version) ||
+                execArgs.Contains(Args::Type::Channel) ||
+                execArgs.Contains(Args::Type::Exact);
+        }
+
+        // Determines whether there are any arguments only used when upgrading a single package,
+        // as opposed to upgrading multiple packages or listing all available upgrades
+        bool HasArgumentsForSinglePackage(Execution::Args& execArgs)
+        {
+            return HasSearchQueryArguments(execArgs) ||
+                execArgs.Contains(Args::Type::Manifest);
+        }
+
+        // Determines whether there are any arguments only used when dealing with multiple packages,
+        // either for upgrading or for listing available upgrades.
+        bool HasArgumentsForMultiplePackages(Execution::Args& execArgs)
+        {
+            return execArgs.Contains(Args::Type::All) ||
+                execArgs.Contains(Args::Type::IncludeUnknown);

```

Comment:
I feel IncludeUnknown is more semantically related to search result filtering. Do we want to apply IncludeUnknown to single query in the future? Or else maybe we have a separate validation to tie IncludeUnknown to multiple packages only (it can only be used for list upgrade or upgrade all) so this fits in HasArgumentsForMultiplePackages?

Reply:
`--include-unknown` as it works right now works for single applications, which is logical to me (because if I have two packages installed with unknown versions but I know only one of them is out of date, I don't want winget to reinstall both of them). 

@lechacon, thoughts?

Reply:
> `--include-unknown` as it works right now works for single applications, which is logical to me

Oh my memory is failing me. I thought it only works for multiple packages. Now I think we should move IncludeUnknown out of HasArgumentsForMultiplePackages then.

Reply:
Resolved in latest commit!